File: Namyata/LongestHarmoniusSubsequence.py
# We define a harmonious array as an array where the difference between its maximum value and its minimum value is exactly 1.
#
# Given an integer array nums, return the length of its longest harmonious subsequence among all its possible subsequences.
#
# A subsequence of array is a sequence that can be derived from the array by deleting some or no elements without changing the order of the remaining elements.

import unittest

def findLHS(nums) -> int:
    freq = {}
    for num in nums:
        freq[num] = freq.get(num, 0) + 1
    max_length = 0
    for num in freq:
        if num + 1 in freq:
            max_length = max(max_length, freq[num] + freq[num + 1])
    return max_length

class TestFindLHS(unittest.TestCase):
    def test_example_case(self):
        nums = [1, 3, 2, 2, 5, 2, 3, 7]
        expected = 5
        self.assertEqual(findLHS(nums), expected)

    def test_empty_list(self):
        nums = []
        expected = 0
        self.assertEqual(findLHS(nums), expected)

    def test_single_element(self):
        nums = [5]
        expected = 0
        self.assertEqual(findLHS(nums), expected)

    def test_no_consecutive_elements(self):
        nums = [1, 2, 3, 4, 5]
        expected = 0
        # expected = 2
        self.assertEqual(findLHS(nums), expected)

    def test_negative_numbers(self):
        nums = [-1, -2, -2, -2, -3, -4, -5]
        expected = 3
        # expected = 4
        self.assertEqual(findLHS(nums), expected)

    def test_large_input(self):
        nums = [10**9] * 10**4 + [0] * 10**4
        expected = 10**4
        # expected = 0
        self.assertEqual(findLHS(nums), expected)

if __name__ == '__main__':
    unittest.main()

# 3 test cases by chatGPT had to be corrected. Uncomment the lines to run the correct version.

File: CTests/longest_harmonius_subsequence.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int findLHS(int* nums, int numsSize) {
    int* freq = (int*)calloc(20001, sizeof(int));
    for (int i = 0; i < numsSize; i++) {
        freq[nums[i] + 10000]++;
    }

    int maxLength = 0;
    for (int i = 0; i < 20000; i++) {
        if (freq[i] > 0 && freq[i + 1] > 0) {
            int length = freq[i] + freq[i + 1];
            if (length > maxLength) {
                maxLength = length;
            }
        }
    }

    free(freq);
    return maxLength;
}

bool testExampleCase() {
    int nums[] = {1, 3, 2, 2, 5, 2, 3, 7};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int expected = 5;
    int result = findLHS(nums, numsSize);
    if (result == expected) {
        printf("Example Case: Passed\n");
        return true;
    } else {
        printf("Example Case: Failed\n");
        return false;
    }
}

bool testEmptyList() {
    int nums[] = {};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int expected = 0;
    int result = findLHS(nums, numsSize);
    if (result == expected) {
        printf("Empty List Case: Passed\n");
        return true;
    } else {
        printf("Empty List Case: Failed\n");
        return false;
    }
}

bool testSingleElement() {
    int nums[] = {5};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int expected = 0;
    int result = findLHS(nums, numsSize);
    if (result == expected) {
        printf("Single Element Case: Passed\n");
        return true;
    } else {
        printf("Single Element Case: Failed\n");
        return false;
    }
}

bool testNoConsecutiveElements() {
    int nums[] = {1, 2, 3, 4, 5};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int expected = 2;
    int result = findLHS(nums, numsSize);
    if (result == expected) {
        printf("No Consecutive Elements Case: Passed\n");
        return true;
    } else {
        printf("No Consecutive Elements Case: Failed\n");
        return false;
    }
}

bool testNegativeNumbers() {
    int nums[] = {-1, -2, -2, -2, -3, -4, -5};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int expected = 4;
    int result = findLHS(nums, numsSize);
    if (result == expected) {
        printf("Negative Numbers Case: Passed\n");
        return true;
    } else {
        printf("Negative Numbers Case: Failed\n");
        return false;
    }
}

bool testLargeInput() {
    int* nums = (int*)malloc(20000 * sizeof(int));
    for (int i = 0; i < 10000; i++) {
        nums[i] = 1000000000;
        nums[i + 10000] = 0;
    }
    int expected = 0;
    int result = findLHS(nums, 20000);
    if (result == expected) {
        printf("Large Input Case: Passed\n");
        return true;
    } else {
        printf("Large Input Case: Failed\n");
        return false;
    }
}

void runTestCases() {
    bool success = true;
    success &= testExampleCase();
    success &= testEmptyList();
    success &= testSingleElement();
    success &= testNoConsecutiveElements();
    success &= testNegativeNumbers();
    success &= testLargeInput();
    
    if (success) {
        printf("All test cases passed!\n");
    } else {
        printf("Some test cases failed!\n");
    }
}

int main() {
    runTestCases();

    return 0;
}
